lib/internal/process/modules.js

```diff
@@ -23,25 +37,81 @@ function initializeImportMetaObject(wrap, meta) {
 
 let loaderResolve;
 exports.loaderPromise = new Promise((resolve, reject) => {
-  loaderResolve = resolve;
+  loaderResolve = v => {
+    resolve(v);
+  }
 });
 
 exports.ESMLoader = undefined;
 
 exports.setup = function() {
   setInitializeImportMetaObjectCallback(initializeImportMetaObject);
 
-  let ESMLoader = new Loader();
+  const RuntimeLoader = new Loader();
   const loaderPromise = (async () => {
-    const userLoader = process.binding('config').userLoader;
-    if (userLoader) {
-      const hooks = await ESMLoader.import(
-        userLoader, getURLFromFilePath(`${process.cwd()}/`).href);
-      ESMLoader = new Loader();
-      ESMLoader.hook(hooks);
-      exports.ESMLoader = ESMLoader;
+    const { userLoaders } = process.binding('config');
+    if (userLoaders) {
+      const BootstrapLoader = new Loader();
+      exports.ESMLoader = BootstrapLoader;
+      let resolve = (url, referrer) => {
+        return require('internal/loader/DefaultResolve')(url, referrer);
+      };
+      let dynamicInstantiate = (url) => {
+        throw new errors.Error('ERR_MISSING_DYNAMIC_INSTANTIATE_HOOK');
+      };
+      for (var i = 0; i < userLoaders.length; i++) {
+        const loaderSpecifier = userLoaders[i];
+        const { default: factory } = await BootstrapLoader.import(
+          loaderSpecifier);
+        const cachedResolve = resolve;
+        const cachedDynamicInstantiate = dynamicInstantiate;
+        const next = factory({
+          __proto__: null,
+          resolve: Object.setPrototypeOf(async (url, referrer) => {
+            const ret = await cachedResolve(url, referrer);
+            return {
+              __proto__: null,
+              url: `${ret.url}`,
+              format: `${ret.format}`,
+            };
+          }, null),
+          dynamicInstantiate: Object.setPrototypeOf(async (url) => {
+            const ret = await cachedDynamicInstantiate(url);
+            return {
+              __proto__: null,
+              exports: ret.exports,
+              execute: ret.execute,
+            };
+          }, null),
+        });
```

Comment:
Can you add some comments about these two functions? For somebody not familiar with loaders, they will be hard to read.